channeld/channeld.c

```diff
@@ -2199,40 +2586,1304 @@ static void handle_unexpected_reestablish(struct peer *peer, const u8 *msg)
+
+	/* Find splice_funding_index */
+	for (int i = 0; i < signed_psbt->num_inputs; i++) {
+		struct wally_tx_input *in = &signed_psbt->tx->inputs[i];
+
+		if (0 != memcmp(in->txhash,
+			       &peer->channel->funding.txid,
+			       sizeof(in->txhash)))
+			continue;
+
+		if (peer->channel->funding.n == in->index) {
+			splice_funding_index = i;
+			break;
+		}
+	}
+
+	if (splice_funding_index == -1)
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Unable to find splice funding tx");
+
+	status_debug("Splice signing tx: %s", tal_hex(tmpctx, linearize_tx(tmpctx, bitcoin_tx)));
+
+	/* Have HSMD sign over the funding tx -> splice tx */
+	msg = towire_hsmd_sign_splice_tx(tmpctx, bitcoin_tx,
+					 &peer->channel->funding_pubkey[REMOTE],
+					 splice_funding_index);
+	msg = hsm_req(tmpctx, take(msg));
+	if (!fromwire_hsmd_sign_tx_reply(msg, &splice_sig))
+		status_failed(STATUS_FAIL_HSM_IO,
+			      "Reading sign_splice_tx reply: %s",
+			      tal_hex(tmpctx, msg));
+
+	/* Set the splice_sig on the splice funding tx psbt */
+	if (!psbt_input_set_signature(signed_psbt, splice_funding_index,
+				      &peer->channel->funding_pubkey[LOCAL],
+				      &splice_sig))
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Unable to set signature internally "
+			      "funding_index: %d "
+			      "my pubkey: %s "
+			      "my signature: %s "
+			      "psbt: %s",
+			      splice_funding_index,
+			      type_to_string(tmpctx, struct pubkey, &peer->channel->funding_pubkey[LOCAL]),
+			      type_to_string(tmpctx, struct bitcoin_signature, &splice_sig),
+			      type_to_string(tmpctx, struct wally_psbt, signed_psbt));
+
+	txsig_tlvs = tlv_txsigs_tlvs_new(tmpctx);
+	der_len = signature_to_der(der, &splice_sig);
+	txsig_tlvs->funding_outpoint_sig = tal_dup_arr(tmpctx, u8, der,
+						       der_len, 0);
+
+	ws = psbt_to_witness_stacks(tmpctx, signed_psbt,
+				    TX_INITIATOR, splice_funding_index);
+	sigout_msg = towire_tx_signatures(tmpctx, &peer->channel_id,
+					  &signed_psbt_txid, ws, txsig_tlvs);
+
+	if (do_i_sign_first(peer, signed_psbt, TX_INITIATOR)) {
+		status_debug("Splice initiator: we sign first");
+		outmsg = towire_channeld_update_inflight(NULL, signed_psbt);
+		wire_sync_write(MASTER_FD, take(outmsg));
+		peer_write(peer->pps, take(sigout_msg));
+	}
+
+	msg = peer_read(tmpctx, peer->pps);
+	if (handle_peer_error(peer->pps, &peer->channel_id, msg))
+		return;
+
+	if (fromwire_peektype(msg) != WIRE_TX_SIGNATURES)
+		peer_failed_warn(peer->pps, &peer->channel_id,
+				"Splicing got incorrect message from peer: %s "
+				"(should be WIRE_TX_SIGNATURES)",
+				peer_wire_name(fromwire_peektype(msg)));
+
+	if (!fromwire_tx_signatures(tmpctx, msg, &cid, &txid,
+				    cast_const3(struct witness_stack ***, &ws),
+				    &txsig_tlvs))
+		peer_failed_warn(peer->pps, &peer->channel_id,
+			    "Splicing bad tx_signatures %s",
+			    tal_hex(msg, msg));
+
+	their_sig.sighash_type = SIGHASH_ALL;
```

Comment:
nit: is this in the spec? we should probs add a BOLT comment?

Reply:
iz in!